File: src/tuner/data/items.py
import string
from pprint import pformat
from typing import Any, Callable
from src.tuner.data.sizing import PG_SIZING
from pydantic import BaseModel, Field

__all__ = ['PG_TUNE_ITEM']

# =============================================================================
# This section is managed by the application
_FLOAT_PRECISION = 4

class PG_TUNE_ITEM(BaseModel):
    key: str = Field(..., description="The key of the sysctl configuration", frozen=True)
    before: Any = Field(..., description="The system information value before tuning", frozen=True)
    after: Any = Field(..., description="The system information value after tuning", frozen=False)
    comment: str | None = Field(None, description="The comment about the tuning process", frozen=True)

    # Custom-reserved variables for developers
    style: str | None = Field(
        default="$1 = '$2'", frozen=True,
        description="The style of the tuning process. This is used to determine the style of the tuning"
    )
    trigger: Any = Field(description="The function that has been used to trigger the tuning for re-try", frozen=True)
    partial_func: Callable | None = (
        Field(default=None, frozen=True,
              description="The partial function to output the result after finish the tuning. This override the "
                          "output format on the `after` field.")
    )
    hardware_scope: tuple[str, PG_SIZING] = (
        Field(frozen=True,
              description="The hardware scope of the tuning. The first value is the hardware type, the second value is "
                          "its associated level (mini, medium, large, mall, bigt, ...)")
    )

    def out(self, include_comment: bool = False, custom_style: str | None = None) -> str:
        texts = []
        if include_comment:
            comment = str(pformat(self.comment)).replace('\n', '\n# ')
            texts.append(f"# {comment}")

        style = custom_style or self.style or "$1 = $2"
        assert "$1" in style and "$2" in style, f"Invalid style configuration: {style} due to missing $1 and $2"
        if '  ' in style:
            style = ' '.join(style.split())
        after: str = self.out_display()
        style = style.replace("$1", self.key).replace("$2", after).strip()
        texts.append(f'\n{style}' if texts else style)
        return ''.join(texts)

    def out_display(self, override_value=None) -> str:
        after = self.after
        if override_value is not None:
            after = override_value

        if self.partial_func is not None:  # This function is used when we have hard-coded the output format already
            after = self.partial_func(after)
        elif isinstance(after, float):
            after = f'{round(after, _FLOAT_PRECISION):.{_FLOAT_PRECISION}f}'
        if not isinstance(after, str):  # Force to be the string for easy text wrap-up
            after = str(after)
        if '.' in after:
            # Remove un-necessary whitespace and trailing zeros to beautify the output if it is a float
            after = after.strip().rstrip('0')
            if after.endswith('.'):  # We have over-stripped the trailing zero, add one 'zero' back
                after = f'{after}0'

        # Wrap the text
        if isinstance(self.after, str) and (' ' in after or any(char in string.punctuation for char in after)):
            after = f"'{after}'"
        return after

    def transform_keyname(self) -> str:
        # Text Transformation: Remove underscores to whitespace and capitalize the first character of each letter
        return ' '.join([x.capitalize() for x in self.key.split('_')])

    def __repr__(self):
        return (f"PG_TUNE_ITEM(key={self.key}, before={self.before}, style={self.style}, trigger={self.trigger}, "
                f"after={self.after}, partial_func={self.partial_func}, hardware_scope={self.hardware_scope})")

File: js/tuner/data/items.js
// -------------------------------------------------------------------------------------
const _FLOAT_PRECISION = 4; // Default float precision for PG_TUNE_ITEM
// The string punctuation characters
const _STRING_PUNCTUATION = String.raw(""!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~"");

class PG_TUNE_ITEM {
    constructor(data) {
        // Required fields
        this.key = data.key;
        this.before = data.before;
        this.after = data.after;
        this.comment = data.comment || null;

        // Custom-reserved variables for developers
        this.style = data.style !== undefined ? data.style : "$1 = '$2'";
        this.trigger = data.trigger;
        this.partial_func = data.partial_func || null;
        this.hardware_scope = data.hardware_scope; // Expected as a tuple [hardware type, sizing value]
    }

    out(output_if_difference_only = false, include_comment = false, custom_style = null) {
        // If output_if_difference_only is true and before equals after, return an empty string.
        if (output_if_difference_only && this.before === this.after) {
            return '';
        }
        let texts = [];

        if (include_comment && this.comment !== null) {
            // Transform the comment by prefixing each line with "# "
            const formattedComment = String(this.comment)
                .split('\n')
                .map(line => `# ${line}`)
                .join('\n');
            texts.push(formattedComment);
        }

        const style = custom_style || this.style || "$1 = $2";
        if (!style.includes("$1") || !style.includes("$2")) {
            throw new Error(`Invalid style configuration: ${style} due to missing $1 and $2`);
        }
        // Remove duplicated spaces if present
        const cleanedStyle = style.replace(/\s\s+/g, ' ');
        const afterDisplay = this.out_display();
        const resultStyle = cleanedStyle.replace("$1", this.key).replace("$2", afterDisplay).trim();

        texts.push(resultStyle);
        return texts.join('');
    }

    out_display(override_value = null) {
        let value = override_value !== null ? override_value : this.after;

        if (this.partial_func && typeof this.partial_func === 'function') {
            value = this.partial_func(value);
        } else if (typeof value === 'number') {
            // Rounding and converting to fixed point string
            value = value.toFixed(_FLOAT_PRECISION);
            // Remove trailing zeros and possible trailing dot
            value = value.replace(/(\.\d*?[1-9])0+$/,'$1').replace(/\.0+$/,'').replace(/\.$/, '.0');
        }
        if (typeof value !== 'string') {
            value = String(value);
        }
        // Trim whitespace if value contains a decimal point and remove trailing zeros
        if (value.includes('.')) {
            value = value.trim().replace(/(\.\d*?)0+$/, '$1');
            if (value.endsWith('.')) {
                value += '0';
            }
        }
        // If the original after value is a string that contains whitespace or punctuation, wrap it in single quotes.
        if (typeof this.after === 'string' &&
            (this.after.includes(' ') || _STRING_PUNCTUATION.split('').some(p => this.after.includes(p)))) {
            value = `'${value}'`;
        }

        if (typeof this.after === 'string' &&
            (this.after.includes(' ') || )) {
            value = `'${value}'`;
        }
        return value;
    }

    transform_keyname() {
        return this.key.split('_').map(part => part.charAt(0).toUpperCase() + part.slice(1)).join(' ');
    }
}

module.exports = { PG_TUNE_ITEM };